### DISCUSSION TITLE
Logtalk to YAML and the k8s space

### DISCUSSION THREAD

**Discussion by mlliarm** [LogtalkDotOrg/logtalk3#194] [General]
Hello,

I've been working lately a lot with YAML files and started wondering if it'd make sense to create a library that would convert Logtalk code to YAML, as a way to generate long YAML files in a more automatic/optimal way. 

I know that there's more or less the opposite (YAML to Prolog parsers).

As it's probably known, [YAML](https://learnxinyminutes.com/docs/yaml/) is a declarative markup language used heavily in the kubernetes (k8s) space for configuration/specification files.

Here is an example of a specification of a typical Kubernetes resource called [PersistentVolumeClaim (PVC)](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#reserving-a-persistentvolume). Pretty much with this the user/admin makes a claim to the kubernetes cluster for a PersistentVolume to be allocated, according to the parameters defined in the PVC yaml file:
 
```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: task-pv-claim
  namespace: default
spec:
  storageClassName: manual
  accessModes:
  - ReadWriteOnce
  resources:
     requests:
        storage: 3Gi
```

Pushing the question even further, do you think it'd make sense to create an operator framework in Logtalk? I've been working lately in creating event-based software (operators) that run on Kubernetes. 

These "operators" work as "guards" in a container inside k8s and make sure that certain things that are specified in the operator have the desired behavior. 

The operator framework what it does is that one defines the Resources (custom or not) that the operator has to operate on, and if it's a custom resource then a custom API gets created that interracts with the k8s API. The framework generates all the necessary YAML files, so this is where the previous project/library could be used.

A somewhat popular operator framework in golang is [kubebuilder](https://book.kubebuilder.io/introduction) if you want to have a look for a static typed OOP C-like language-based framework.

But since Logtalk supports event-based programming, I believe it should be possible to create an operator framework in Logtalk, and also that perhaps it would be more productive/efficient to work with it than with a framework based on golang.

Sharing here after the gentle suggestion of Paulo.

Feel free to share  your thoughts, comments etc.

Michail

**Comment by pmoura** [comment]
Not familiar with Kubernetes. But a quick glance at some of the linked documentation suggests that we could have a set of objects representing e.g. the different types of resources (such as PVC in your example) with a public predicate that would take the configurable parameter values and generate the corresponding YAML fragment.

Given the number of tools that are available to work with Kubernetes, it would be interesting to know of any *pain points* that could be solved or at least alleviated using a more declarative, logic programming approach.

**Reply by mlliarm** [reply]
Thanks for the reply Paulo.

At this moment I'm not aware yet of any *pain points* related to k8s. I know that load-balancing can be tough but then again I think that the system itself takes care of it automatically.

Nonetheless, I'm sure I'll find some optimization or search problems down the road that could be solved using a logic programming approach much easier than with a procedural/OOP programming language.

I'll update this post soon when I find at least one of such problems.

ps: I think a troubleshooting oracle built in a logic programming language could be helpful. Asume you'd feed it the current state of the system (in YAML or other formats) and the error message(s) and it'd give you the right direction(s) into bringing it in a working state. Troubleshooting distributed systems with many components is a pain and it'd be nice for the system administrator to have an ally when trying to find out what went wrong and why.
